src/m365/pa/commands/app/app-permission-ensure.ts

```diff
@@ -0,0 +1,207 @@
+import PowerAppsCommand from '../../../base/PowerAppsCommand';
+import commands from '../../commands';
+import { aadUser } from '../../../../utils/aadUser';
+import { accessToken } from '../../../../utils/accessToken';
+import Auth from '../../../../Auth';
+
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  appName: string;
+  roleName: string;
+  userId?: string;
+  userName?: string;
+  groupId?: string;
+  groupName?: string;
+  tenant?: boolean;
+  environmentName?: string;
+  sendInvitationMail?: boolean;
+  asAdmin?: boolean;
+}
+
+class PaAppPermissionEnsureCommand extends PowerAppsCommand {
+  private static readonly roleNames = ['CanEdit', 'CanView'];
+  public get name(): string {
+    return commands.APP_PERMISSION_ENSURE;
+  }
+
+  public get description(): string {
+    return 'Assigns/updates permissions to a Power Apps app';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        userId: typeof args.options.userId !== 'undefined',
+        userName: typeof args.options.userName !== 'undefined',
+        groupId: typeof args.options.groupId !== 'undefined',
+        groupName: typeof args.options.groupName !== 'undefined',
+        tenant: !!args.options.tenant,
+        environmentName: typeof args.options.environmentName !== 'undefined',
+        sendInvitationMail: !!args.options.sendInvitationMail,
+        asAdmin: !!args.options.asAdmin
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '--appName <appName>'
+      },
+      {
+        option: '--roleName <roleName>',
+        autocomplete: PaAppPermissionEnsureCommand.roleNames
+      },
+      {
+        option: '--userId [userId]'
+      },
+      {
+        option: '--userName [userName]'
+      },
+      {
+        option: '--groupId [groupId]'
+      },
+      {
+        option: '--groupName [groupName]'
+      },
+      {
+        option: '--tenant'
+      },
+      {
+        option: '-e, --environmentName [environmentName]'
+      },
+      {
+        option: '--sendInvitationMail'
+      },
+      {
+        option: '--asAdmin'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.appName)) {
+          return `${args.options.appName} is not a valid GUID`;
+        }
+
+        if (args.options.userId && !validation.isValidGuid(args.options.userId)) {
+          return `${args.options.userId} is not a valid GUID`;
+        }
+
+        if (args.options.groupId && !validation.isValidGuid(args.options.groupId)) {
+          return `${args.options.groupId} is not a valid GUID`;
+        }
```

Comment:
We have a lot of commands that throw an error like this, but I'm becoming more and more of a fan of explicitly logging the option name in the error. In that case, it's immediately clear to the user which option is blocking him. This can be tough when using variables.
Therefore let's use `${args.options.groupId} is not a valid GUID for groupId` for example.